src/Rib/Extra/OpenGraph.hs

```diff
@@ -19,31 +24,76 @@ import qualified Text.URI as URI
 data OpenGraph
   = OpenGraph
       { _openGraph_title :: Text,
-        _openGraph_author :: Text,
+        _openGraph_url :: URI.URI,
+        _openGraph_author :: Maybe Text,
         _openGraph_description :: Maybe Text,
         _openGraph_siteName :: Text,
-        _openGraph_type :: Text, -- TODO: ADT (different types with their own properties)
+        _openGraph_type :: Maybe OGType,
         _openGraph_image :: Maybe URI.URI
       }
   deriving (Eq, Show)
 
 instance ToHtml OpenGraph where
   toHtmlRaw = toHtml
   toHtml OpenGraph {..} = do
-    meta' "author" _openGraph_author
+    meta' "author" `mapM_` _openGraph_author
     meta' "description" `mapM_` _openGraph_description
+    requireAbsolute "OGP URL" (\uri -> link_ [rel_ "canonical", href_ uri]) _openGraph_url
     metaOg "title" _openGraph_title
     metaOg "site_name" _openGraph_siteName
-    metaOg "type" _openGraph_type
-    whenJust _openGraph_image $ \uri -> do
-      if isJust (URI.uriScheme uri)
-        then metaOg "image" $ URI.render uri
-        else error $ "OGP image URL must be absolute. This URI is not: " <> URI.render uri
+    toHtml `mapM_` _openGraph_type
+    requireAbsolute "OGP image URL" (metaOg "image") `mapM_` _openGraph_image
     where
-      -- Open graph meta element
-      metaOg k v =
-        meta_
-          [ makeAttribute "property" $ "og:" <> k,
-            content_ v
-          ]
       meta' k v = meta_ [name_ k, content_ v]
+      requireAbsolute description f uri =
+        if isJust (URI.uriScheme uri)
+          then f $ URI.render uri
+          else error $ description <> " must be absolute. this URI is not: " <> URI.render uri
+
+-- TODO: Remaining ADT values & sub-fields
+data OGType
+  = OGType_Article OGArticle
+  deriving (Eq, Show)
+
+instance ToHtml OGType where
+  toHtmlRaw = toHtml
+  toHtml = \case
+    OGType_Article article -> do
+      metaOg "type" "article"
+      toHtml article
+
+
+-- TODO: _article_profile :: [Profile]
+data OGArticle
+  = OGArticle
+    { _article_section :: Maybe Text
+    , _article_modifiedTime :: Maybe UTCTime
+    , _article_publishedTime :: Maybe UTCTime
+    , _article_expirationTime :: Maybe UTCTime
+    , _article_tag :: [Text]
```

Comment:
Naming here should be of the form `_ogArticle_section` to be consistent. Either that, or you can simply rename `OGArticle` to `Article`. I have a slight preference for the latter.

Reply:
I went w/ OGArticle because I thought maybe someone's site might have their own Article type or constructor.